### PR TITLE
ESQL: Rework sequence for analyzing queries

### LINKED ISSUE #107953: ESQL: Refactor top-level query planning
ESQL's query planning currently uses multiple nested ActionListeners (see [entrypoint here](https://github.com/elastic/elasticsearch/blob/3ed42f38c3363b5108d0ed07c49af186d438694e/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java#L125). Because of this, the code needs to be read last-to-first, starting with the plan optimization (last step), going deeper until the initial planning step (first step).

The ActionListeners seem to only contain procedural code, so this can probably be rewritten as simple procedural code: parse -> plan -> optimizations, resp. the number of action listener layers can likely be reduced.

### PR DESCRIPTION
**Pull request by nik9000**
In service of the incoming INLINESTATS this flips the ordering of analysis. Previously we made the entire sequence of analyze, optimize, convert to physical plan, and optimize in a single async sequence in `EsqlSession`. This flips it so `analyze` comes first in it's own async sequence and then runs the remaining stuff in a separate sequence. That's nice for INLINESTATS where we want to analyze one time, and then run many runs of the extra sequence.

While we're here, we also take that sequence call it directly from the CsvTests. That works well because that final sequence is exactly what CsvTests have to do. They "analyze" totally differently, but they run the final sequence in the same way.

Closes #107953

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java [hunk 1]
```diff
@@ -155,6 +150,7 @@ public void analyzedPlan(LogicalPlan parsed, ActionListener<LogicalPlan> listene
         preAnalyze(parsed, (indices, policies) -> {
             Analyzer analyzer = new Analyzer(new AnalyzerContext(configuration, functionRegistry, indices, policies), verifier);
             var plan = analyzer.analyze(parsed);
+            plan.setAnalyzed();
```
Comment (nik9000) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:153]: Now that analyzing runs in a separate sequence I've started setting this flag and checking it on the other side as an extra check.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/CsvTests.java [hunk 2]
```diff
@@ -381,90 +380,43 @@ private static TestPhysicalOperationProviders testOperationProviders(CsvTestsDat
...
+            null,
+            null,
+            null,
+            functionRegistry,
+            new LogicalPlanOptimizer(new LogicalOptimizerContext(configuration)),
+            mapper,
+            TEST_VERIFIER
         );
-        //
-        // Keep in sync with ComputeService#execute
```
Comment (nik9000) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/CsvTests.java:409]: The stuff in EsqlSession is now just called directly.

**Review by dnhatn — APPROVED** [review]
LGTM. Thanks Nik!

**Review by ChrisHegarty — APPROVED** [review]
LGTM

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/CsvTests.java (hunk 2 above)
Comment (alex-spies) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/CsvTests.java:409]: This is nice, but I'd still preserve this comment and place it on top of `runPhase` in line 465; the stuff in there still needs to be close to how the real compute engine runs.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java [hunk 3]
```diff
@@ -315,28 +311,41 @@ private static Set<String> subfields(Set<String> names) {
         return names.stream().filter(name -> name.endsWith(WILDCARD) == false).map(name -> name + ".*").collect(Collectors.toSet());
     }
 
-    public void optimizedPlan(LogicalPlan logicalPlan, ActionListener<LogicalPlan> listener) {
-        analyzedPlan(logicalPlan, map(listener, p -> {
-            var plan = logicalPlanOptimizer.optimize(p);
-            LOGGER.debug("Optimized logicalPlan plan:\n{}", plan);
-            return plan;
-        }));
+    private PhysicalPlan logicalPlanToPhysicalPlan(LogicalPlan logicalPlan, EsqlQueryRequest request) {
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:314]: Neat, thanks for factoring away the nested action listeners.

This addresses https://github.com/elastic/elasticsearch/issues/107953.

**Review by alex-spies — APPROVED** [review]
LGTM, thanks Nik!

**Pull request merged**